src/Microsoft.Azure.SignalR.Common/ClientInvocation/ClientResultsManager.cs

```diff
@@ -0,0 +1,323 @@
+        {
+            if (_serviceMappingMessages.TryRemove(instanceId, out var serviceMappingMessage))
+            {
+                if (_pendingInvocations.TryRemove(serviceMappingMessage.InvocationId, out var item))
+                {
+                    var message = new CompletionMessage(serviceMappingMessage.InvocationId, "mapping service offline", null, true);
+                    item.Complete(item.Tcs, message);
+                }
+            }
+        }
+
+        public Task<T> AddRoutedInvocation<T>(string connectionId, string invocationId, string callerServerId, CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSourceWithCancellation<T>(this, connectionId, invocationId, cancellationToken);
+            var result = _routedInvocations.TryAdd(invocationId, (typeof(T), connectionId, callerServerId, tcs, static (state, completionMessage) =>
+            {
+                var tcs = (TaskCompletionSourceWithCancellation<T>)state;
+                if (completionMessage.HasResult)
+                {
+                    tcs.SetResult((T)completionMessage.Result);
+                }
+                else
+                {
+                    tcs.SetException(new Exception(completionMessage.Error));
+                }
+            }
+            ));
+            Debug.Assert(result);
+
+            tcs.RegisterCancellation();
+
+            return tcs.Task;
+        }
+
+        public Task<T> AddInvocation<T>(string connectionId, string invocationId, CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSourceWithCancellation<T>(this, connectionId, invocationId, cancellationToken);
+            var result = _pendingInvocations.TryAdd(invocationId, (typeof(T), connectionId, tcs, static (state, completionMessage) =>
+            {
+                var tcs = (TaskCompletionSourceWithCancellation<T>)state;
+                if (completionMessage.HasResult)
+                {
+                    tcs.SetResult((T)completionMessage.Result);
+                }
+                else
+                {
+                    tcs.SetException(new Exception(completionMessage.Error));
+                }
+            }
+            ));
+            Debug.Assert(result);
+
+            tcs.RegisterCancellation();
+
+            return tcs.Task;
+        }
+
+        public void AddInvocation(string invocationId, (Type Type, string ConnectionId, object Tcs, Action<object, CompletionMessage> Complete) invocationInfo)
+        {
+            var result = _pendingInvocations.TryAdd(invocationId, invocationInfo);
+            Debug.Assert(result);
+        }
+
+        public void TryCompleteResult(string connectionId, string protocol, ReadOnlySequence<byte> message)
+        {
+            var proto = _hubProtocolResolver.GetProtocol(protocol, new string[] { protocol });
+            if (proto == null)
+            {
+                throw new InvalidOperationException();
```

Comment:
Add detail information. Shall we send a `ErrorCompletionMessage` back to service including some detail information here detected by route server. So original caller would be aware what's wrong.

Reply:
Looks this is final server. So no need to tell service again. Simply fulfill the exception is good.

Reply:
Fixed in #1684